sls-versions/src/main/java/com/palantir/sls/versions/HandRolledMatcherParser.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * (c) Copyright 2021 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.sls.versions;
+
+import java.util.Optional;
+import java.util.OptionalInt;
+
+/**
+ * A hand-rolled implementation of {@link RegexMatcherParser}.
+ *
+ * Uses two-element int[] arrays as the 'result' type to represent a successful/failed parse, where result[0] always
+ * contains the index into the string we're looking at, and result[1] contains the value returned by the parser
+ * combinator.
+ */
+final class HandRolledMatcherParser {
+
+    private static final int MAGIC_X_NUMBER = -1;
+
+    public static Optional<SlsVersionMatcher> safeValueOf(String string) {
+        OptionalInt major = OptionalInt.empty();
+        OptionalInt minor = OptionalInt.empty();
+        OptionalInt patch = OptionalInt.empty();
+
+        int[] result = new int[] {0};
+
+        // major
+        result = numberOrX(string, result[0]);
```

Comment:
Dead array allocation. I'm not sure why we reuse the same reference for subsequent arrays, it might. be easier to follow if we used `int[] major =...`. That said, I'm not sure why we're using arrays when `long` is a great `int[2]`.
```suggestion
        // major
        int[] result = numberOrX(string, 0);
```